x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,152 @@
+
+    public static final CoordinatorProxyAction INSTANCE = new CoordinatorProxyAction();
+    public static final String NAME = "indices:data/read/xpack/enrich/coordinate_lookups";
+
+    private CoordinatorProxyAction() {
+        super(NAME, SearchResponse::new);
+    }
+
+    public static class TransportAction extends HandledTransportAction<SearchRequest, SearchResponse> {
+
+        private final Coordinator coordinator;
+
+        @Inject
+        public TransportAction(TransportService transportService, ActionFilters actionFilters, Coordinator coordinator) {
+            super(NAME, transportService, actionFilters, (Writeable.Reader<SearchRequest>) SearchRequest::new);
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> listener) {
+            assert Thread.currentThread().getName().contains(ThreadPool.Names.WRITE);
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
+
+        final Client client;
+        final int maxLookupsPerRequest;
+        final int maxNumberOfConcurrentRequests;
+        final BlockingQueue<Item> queue = new LinkedBlockingQueue<>();
+        final AtomicInteger numberOfOutstandingRequests = new AtomicInteger(0);
+
+        public Coordinator(Client client, Settings settings) {
+            this(client,
+                EnrichPlugin.COORDINATOR_PROXY_MAX_LOOKUPS_PER_REQUEST.get(settings),
+                EnrichPlugin.COORDINATOR_PROXY_MAX_CONCURRENT_REQUESTS.get(settings));
+        }
+
+        Coordinator(Client client, int maxLookupsPerRequest, int maxNumberOfConcurrentRequests) {
+            this.client = client;
+            this.maxLookupsPerRequest = maxLookupsPerRequest;
+            this.maxNumberOfConcurrentRequests = maxNumberOfConcurrentRequests;
+        }
+
+        void schedule(SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
+            // Use put(...), because if queue is full then this method will wait until a free slot becomes available
+            // The calling thread here is a write thread (write tp is used by ingest) and
+            // this will great natural back pressure from the enrich processor.
+            // If there are no write threads available then write requests with ingestion will fail with 429 error code.
+            try {
+                queue.put(new Item(searchRequest, listener));
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException("unable to add item to queue", e);
+            }
+            coordinateLookups();
+        }
+
+        synchronized void coordinateLookups() {
+            while (queue.isEmpty() == false &&
+                numberOfOutstandingRequests.get() <= maxNumberOfConcurrentRequests) {
```

Comment:
I think it should be `<` rather than `<=` otherwise we could have `maxNumberOfConcurrentRequests+1` concurrent requests?